File: internal/usecase/ports/inbox.go
package ports

import (
	"context"
	"time"

	"github.com/KKloudTarus/synapse-ce/internal/domain/notification"
	"github.com/KKloudTarus/synapse-ce/internal/domain/shared"
)

type InboxItem struct {
	ID        shared.ID  `json:"id"`
	EventID   shared.ID  `json:"event_id"`
	EventType string     `json:"event_type"`
	Title     string     `json:"title"`
	Summary   string     `json:"summary"`
	LinkPath  string     `json:"link_path"`
	CreatedAt time.Time  `json:"created_at"`
	ReadAt    *time.Time `json:"read_at,omitempty"`
}

type InboxPage struct {
	Items []InboxItem `json:"items"`
	Next  string      `json:"next,omitempty"`
}

type InboxPreference struct {
	EventType string                  `json:"event_type"`
	Channel   string                  `json:"channel"`
	State     notification.Preference `json:"state"`
	Revision  int                     `json:"revision"`
	Mandatory bool                    `json:"mandatory"`
	Available bool                    `json:"available"`
	Reason    string                  `json:"reason,omitempty"`
}

type InboxStore interface {
	ListInbox(context.Context, shared.ID, shared.ID, time.Time, shared.ID, bool, int) ([]InboxItem, error)
	UnreadInbox(context.Context, shared.ID, shared.ID) (int, error)
	MarkInboxRead(context.Context, shared.ID, shared.ID, shared.ID, time.Time) error
	MarkInboxAllRead(context.Context, shared.ID, shared.ID, time.Time) error
	ListInboxPreferences(context.Context, shared.ID, shared.ID) ([]InboxPreference, error)
	SaveInboxPreference(context.Context, shared.ID, shared.ID, notification.EventType, string, notification.Preference, int, time.Time) (InboxPreference, error)
	LoadPersonalMail(context.Context, shared.ID, shared.ID, shared.ID, shared.ID, int) (string, string, string, bool, error)
}

type PersonalNoticeMailer interface {
	SendPersonalNotice(context.Context, string, string, string, shared.ID) NotificationSendResult
}

File: web/src/lib/api/inbox.ts
import { req } from './client'

export type InboxItem = {
  id: string
  event_id: string
  event_type: string
  title: string
  summary: string
  link_path: string
  created_at: string
  read_at?: string
}

export type InboxPage = { items: InboxItem[]; next?: string }

export type InboxPreference = {
  event_type: string
  channel: string
  state: 'inherit' | 'enabled' | 'disabled'
  revision: number
  mandatory: boolean
  available: boolean
  reason?: string
}

export const inboxApi = {
  inboxPage: (cursor?: string, unread = false, signal?: AbortSignal): Promise<InboxPage> => {
    const params = new URLSearchParams({ limit: '25' })
    if (cursor) params.set('cursor', cursor)
    if (unread) params.set('unread', 'true')
    return req(`/me/inbox?${params}`, signal ? { signal } : undefined)
  },
  inboxUnread: (signal?: AbortSignal): Promise<{ unread: number }> => req('/me/inbox/unread', signal ? { signal } : undefined),
  markInboxRead: (id: string): Promise<void> => req(`/me/inbox/${encodeURIComponent(id)}/read`, { method: 'POST' }),
  markInboxAllRead: (): Promise<void> => req('/me/inbox/read', { method: 'POST' }),
  inboxPreferences: (signal?: AbortSignal): Promise<{ items: InboxPreference[] }> => req('/me/notification-preferences', signal ? { signal } : undefined),
  saveInboxPreference: (input: Pick<InboxPreference, 'event_type' | 'channel' | 'state' | 'revision'>): Promise<InboxPreference> =>
    req('/me/notification-preferences', { method: 'PUT', body: JSON.stringify(input) }),
}
